deployment/hasura/migrations/AerieMerlin/2_drop_profile_topic_fks/up.sql

```diff
@@ -0,0 +1,290 @@
+
+comment on procedure span_add_foreign_key_to_partition is e''
+  'Creates a self-referencing foreign key on a particular partition of the span table. This should be called'
+  'on every partition as soon as it is created';
+
+-- dataset
+create or replace function delete_dataset_cascade()
+  returns trigger
+  security definer
+  language plpgsql as $$begin
+  delete from span where span.dataset_id = old.id;
+  return old;
+end$$;
+
+create or replace function delete_dataset_cascade()
+  returns trigger
+  security definer
+  language plpgsql as $$begin
+  delete from span where span.dataset_id = old.id;
+  return old;
+end$$;
+
+create trigger delete_dataset_trigger
+  after delete on dataset
+  for each row
+execute function delete_dataset_cascade();
+
+comment on trigger delete_dataset_trigger on dataset is e''
+  'Trigger to simulate an ON DELETE CASCADE foreign key constraint between span and dataset. The reason to'
+  'implement this as a trigger is that this single trigger can cascade deletes to any partitions of span.'
+  'If we used a foreign key, every new partition of span would need to add a new trigger to the dataset'
+  'table - which requires acquiring a lock that conflicts with concurrent inserts. In order to allow adding new'
+  'partitions concurrently with inserts to referenced tables, we have chosen to forego foreign keys from partitions'
+  'to other tables in favor of these hand-written triggers';
+
+create or replace function allocate_dataset_partitions(dataset_id integer)
+  returns dataset
+  security definer
+  language plpgsql as $$
+declare
+  dataset_ref dataset;
+begin
+  select * from dataset where id = dataset_id into dataset_ref;
+  if dataset_id is null
+  then
+    raise exception 'Cannot allocate partitions for non-existent dataset id %', dataset_id;
+  end if;
+
+  execute 'create table profile_segment_' || dataset_id || ' (
+    like profile_segment including defaults including constraints
+  );';
+  execute 'alter table profile_segment
+    attach partition profile_segment_' || dataset_id || ' for values in ('|| dataset_id ||');';
+
+  execute 'create table event_' || dataset_id || ' (
+      like event including defaults including constraints
+    );';
+  execute 'alter table event
+    attach partition event_' || dataset_id || ' for values in (' || dataset_id || ');';
+
+  execute 'create table span_' || dataset_id || ' (
+       like span including defaults including constraints
+    );';
+  execute 'alter table span
+    attach partition span_' || dataset_id || ' for values in (' || dataset_id || ');';
+
+  -- Create a self-referencing foreign key on the span partition table. We avoid referring to the top level span table
+  -- in order to avoid lock contention with concurrent inserts
+  execute 'alter table span_' || dataset_id || ' add constraint span_has_parent_span
+    foreign key (dataset_id, parent_id)
+    references span_' || dataset_id || '
+    on update cascade
+    on delete cascade;';
```

Comment:
This part of the method differs here from the version in dataset.sql

There, this part is: `call span_add_foreign_key_to_partition('span_' || dataset_id);`

Reply:
Good catch - updated :+1: